go/src/net/ConnectionPool.go

```diff
@@ -0,0 +1,207 @@
+			if ele.Value.(*Connection).SeverAddress == curSession.connection.SeverAddress {
+				ele.Value.(*Connection).SeverAddress.IsAvaliable = false
+			}
+		}
+		// Mark the current host as non-avaliable
+		curSession.connection.SeverAddress.IsAvaliable = false
+		// Return an idle session that is different from the current one
+		for ele := pool.idleConnectionQueue.Front(); ele != nil; ele = ele.Next() {
+			if ele.Value.(*Connection).SeverAddress != curSession.connection.SeverAddress &&
+				ele.Value.(*Connection).SeverAddress.IsAvaliable == true {
+				return ele.Value.(*Connection), nil
+			}
+		}
+		noAvaliableConnectionErr := errors.New("Failed to reconnect: no avaliable connection to a different host can be found")
+
+		return nil, noAvaliableConnectionErr
+	}
+
+	// Create a new connection if there is no idle connection and total connection < pool max size
+	totalConn := pool.idleConnectionQueue.Len() + pool.activeConnectionQueue.Len()
+	// TODO: use load balencer to decide the host to connect
+	var severAddress *data.HostAddress
+	severAddress = nil
+	// Check if there is valid host in the pool
+	if totalConn < pool.conf.MaxConnPoolSize {
+		for _, host := range pool.addresses {
+			if host.IsAvaliable == true {
+				severAddress = host
+			}
+		}
+		// No valid host in the pool
+		if severAddress == nil {
+			noAvaliableHost := errors.New("Failed to reconnect: no avaliable host in the connection pool")
+			if noAvaliableHost != nil {
+				fmt.Print(noAvaliableHost)
+			}
+			return nil, noAvaliableHost
+		}
+		newConn := NewConnection(*severAddress)
+		newConn.inuse = false
+
+		// Open connection to host
+		err := newConn.Open(newConn.SeverAddress, *pool.conf)
+		if err != nil {
+			log.Printf("Failed to open connection, error: %s \n", err.Error())
+			return nil, err
+		}
+
+		// Add new connection into idle queue
+		pool.idleConnectionQueue.PushBack(newConn)
+		return newConn, nil
+	}
+	// TODO: If no idle avaliable, wait for timeout and reconnect
+
+	// If no idle avaliable and the number of total connection reaches the max pool size, return error/wait for timeout
+	noAvaliableConnectionErr := errors.New("Failed to reconnect: no avaliable connection in the connection pool")
+	if noAvaliableConnectionErr != nil {
+		fmt.Print(noAvaliableConnectionErr)
+	}
+	return nil, noAvaliableConnectionErr
+}
+
+// Close all connection
+func (pool *ConnectionPool) Close() error {
+	for conn := pool.idleConnectionQueue.Front(); conn != nil; conn = conn.Next() {
+		err := conn.Value.(*Connection).Close()
+		if err != nil {
+			log.Printf("Failed to close connection, error: %s", err.Error())
+			return err
+		}
+	}
+	for conn := pool.activeConnectionQueue.Front(); conn != nil; conn = conn.Next() {
+		err := conn.Value.(*Connection).Close()
+		if err != nil {
+			log.Printf("Failed to close connection, error: %s", err.Error())
+			return err
+		}
+	}
+	return nil
+}
+
+func (rsp RespData) String() string {
```

Comment:
What's the use of this?

Reply:
This function forms error messages for resp, but is no long used